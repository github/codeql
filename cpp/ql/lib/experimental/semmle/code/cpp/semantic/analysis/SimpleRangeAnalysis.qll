/**
 * Wrapper for the semantic range analysis library that mimics the
 * interface of the simple range analysis library.
 */

private import cpp
private import semmle.code.cpp.ir.IR
private import experimental.semmle.code.cpp.semantic.SemanticBound
private import experimental.semmle.code.cpp.semantic.SemanticExprSpecific
private import RangeAnalysis

/**
 * Gets the lower bound of the expression.
 *
 * Note: expressions in C/C++ are often implicitly or explicitly cast to a
 * different result type. Such casts can cause the value of the expression
 * to overflow or to be truncated. This predicate computes the lower bound
 * of the expression without including the effect of the casts. To compute
 * the lower bound of the expression after all the casts have been applied,
 * call `lowerBound` like this:
 *
 *    `lowerBound(expr.getFullyConverted())`
 */
float lowerBound(Expr expr) {
  exists(Instruction i, SemBound b | i.getAst() = expr and b instanceof SemZeroBound |
    semBounded(getSemanticExpr(i), b, result, false, _)
  )
}

/**
 * Gets the upper bound of the expression.
 *
 * Note: expressions in C/C++ are often implicitly or explicitly cast to a
 * different result type. Such casts can cause the value of the expression
 * to overflow or to be truncated. This predicate computes the upper bound
 * of the expression without including the effect of the casts. To compute
 * the upper bound of the expression after all the casts have been applied,
 * call `upperBound` like this:
 *
 *    `upperBound(expr.getFullyConverted())`
 */
float upperBound(Expr expr) {
  exists(Instruction i, SemBound b | i.getAst() = expr and b instanceof SemZeroBound |
    semBounded(getSemanticExpr(i), b, result, true, _)
  )
}

/**
 * Holds if the upper bound of `expr` may have been widened. This means the
 * upper bound is in practice likely to be overly wide.
 */
predicate upperBoundMayBeWidened(Expr e) { none() }

/**
 * Holds if `expr` has a provably empty range. For example:
 *
 *   10 < expr and expr < 5
 *
 * The range of an expression can only be empty if it can never be
 * executed. For example:
 *
 * ```cpp
 * if (10 < x) {
 *   if (x < 5) {
 *     // Unreachable code
 *     return x; // x has an empty range: 10 < x && x < 5
 *   }
 * }
 * ```
 */
predicate exprWithEmptyRange(Expr expr) { lowerBound(expr) > upperBound(expr) }

/** Holds if the definition might overflow negatively. */
predicate defMightOverflowNegatively(RangeSsaDefinition def, StackVariable v) { none() }

/** Holds if the definition might overflow positively. */
predicate defMightOverflowPositively(RangeSsaDefinition def, StackVariable v) { none() }

/**
 * Holds if the definition might overflow (either positively or
 * negatively).
 */
predicate defMightOverflow(RangeSsaDefinition def, StackVariable v) {
  defMightOverflowNegatively(def, v) or
  defMightOverflowPositively(def, v)
}

/**
 * Holds if the expression might overflow negatively. This predicate
 * does not consider the possibility that the expression might overflow
 * due to a conversion.
 */
predicate exprMightOverflowNegatively(Expr expr) { none() }

/**
 * Holds if the expression might overflow negatively. Conversions
 * are also taken into account. For example the expression
 * `(int16)(x+y)` might overflow due to the `(int16)` cast, rather than
 * due to the addition.
 */
predicate convertedExprMightOverflowNegatively(Expr expr) {
  exprMightOverflowNegatively(expr) or
  convertedExprMightOverflowNegatively(expr.getConversion())
}

/**
 * Holds if the expression might overflow positively. This predicate
 * does not consider the possibility that the expression might overflow
 * due to a conversion.
 */
predicate exprMightOverflowPositively(Expr expr) { none() }

/**
 * Holds if the expression might overflow positively. Conversions
 * are also taken into account. For example the expression
 * `(int16)(x+y)` might overflow due to the `(int16)` cast, rather than
 * due to the addition.
 */
predicate convertedExprMightOverflowPositively(Expr expr) {
  exprMightOverflowPositively(expr) or
  convertedExprMightOverflowPositively(expr.getConversion())
}

/**
 * Holds if the expression might overflow (either positively or
 * negatively). The possibility that the expression might overflow
 * due to an implicit or explicit cast is also considered.
 */
predicate convertedExprMightOverflow(Expr expr) {
  convertedExprMightOverflowNegatively(expr) or
  convertedExprMightOverflowPositively(expr)
}
