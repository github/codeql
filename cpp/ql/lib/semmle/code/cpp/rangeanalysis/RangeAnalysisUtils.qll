import cpp

/**
 * Describes whether a relation is 'strict' (that is, a `<` or `>`
 * relation) or 'non-strict' (a `<=` or `>=` relation).
 */
newtype RelationStrictness =
  /**
   * Represents that a relation is 'strict' (that is, a `<` or `>` relation).
   */
  Strict() or
  /**
   * Represents that a relation is 'non-strict' (that is, a `<=` or `>=` relation)
   */
  Nonstrict()

/**
 * Describes whether a relation is 'greater' (that is, a `>` or `>=`
 * relation) or 'lesser' (a `<` or `<=` relation).
 */
newtype RelationDirection =
  /**
   * Represents that a relation is 'greater' (that is, a `>` or `>=` relation).
   */
  Greater() or
  /**
   * Represents that a relation is 'lesser' (that is, a `<` or `<=` relation).
   */
  Lesser()

private RelationStrictness negateStrictness(RelationStrictness strict) {
  strict = Strict() and result = Nonstrict()
  or
  strict = Nonstrict() and result = Strict()
}

private RelationDirection negateDirection(RelationDirection dir) {
  dir = Greater() and result = Lesser()
  or
  dir = Lesser() and result = Greater()
}

/**
 * Holds if `dir` is `Greater` (that is, a `>` or `>=` relation)
 */
boolean directionIsGreater(RelationDirection dir) {
  dir = Greater() and result = true
  or
  dir = Lesser() and result = false
}

/**
 * Holds if `dir` is `Lesser` (that is, a `<` or `<=` relation)
 */
boolean directionIsLesser(RelationDirection dir) {
  dir = Greater() and result = false
  or
  dir = Lesser() and result = true
}

/**
 * Holds if `rel` is a relational operation (`<`, `>`, `<=` or `>=`)
 * with fully-converted children `lhs` and `rhs`, described by
 * `dir` and `strict`.
 *
 * For example, if `rel` is `x < 5` then
 * `relOp(rel, x, 5, Lesser(), Strict())` holds.
 */
private predicate relOp(
  RelationalOperation rel, Expr lhs, Expr rhs, RelationDirection dir, RelationStrictness strict
) {
  lhs = rel.getLeftOperand().getFullyConverted() and
  rhs = rel.getRightOperand().getFullyConverted() and
  (
    rel instanceof LTExpr and dir = Lesser() and strict = Strict()
    or
    rel instanceof LEExpr and dir = Lesser() and strict = Nonstrict()
    or
    rel instanceof GTExpr and dir = Greater() and strict = Strict()
    or
    rel instanceof GEExpr and dir = Greater() and strict = Nonstrict()
  )
}

/**
 * Holds if `rel` is a relational operation (`<`, `>`, `<=` or `>=`)
 * with fully-converted children `a` and `b`, described by `dir` and `strict`.
 *
 * This allows for the relation to be either as written, or with its
 * arguments reversed; for example, if `rel` is `x < 5` then both
 * `relOpWithSwap(rel, x, 5, Lesser(), Strict())` and
 * `relOpWithSwap(rel, 5, x, Greater(), Strict())` hold.
 */
predicate relOpWithSwap(
  RelationalOperation rel, Expr a, Expr b, RelationDirection dir, RelationStrictness strict
) {
  relOp(rel, a, b, dir, strict) or
  relOp(rel, b, a, negateDirection(dir), strict)
}

/**
 * Holds if `rel` is a comparison operation (`<`, `>`, `<=` or `>=`)
 * with fully-converted children `a` and `b`, described by `dir` and `strict`, with
 * result `branch`.
 *
 * This allows for the relation to be either as written, or with its
 * arguments reversed; for example, if `rel` is `x < 5` then
 * `relOpWithSwapAndNegate(rel, x, 5, Lesser(), Strict(), true)`,
 * `relOpWithSwapAndNegate(rel, 5, x, Greater(), Strict(), true)`,
 * `relOpWithSwapAndNegate(rel, x, 5, Greater(), Nonstrict(), false)` and
 * `relOpWithSwapAndNegate(rel, 5, x, Lesser(), Nonstrict(), false)` hold.
 */
predicate relOpWithSwapAndNegate(
  RelationalOperation rel, Expr a, Expr b, RelationDirection dir, RelationStrictness strict,
  boolean branch
) {
  relOpWithSwap(rel, a, b, dir, strict) and branch = true
  or
  relOpWithSwap(rel, a, b, negateDirection(dir), negateStrictness(strict)) and
  branch = false
}

/**
 * Holds if `cmp` is an equality operation (`==` or `!=`) with  fully-converted
 * children `lhs` and `rhs`, and `isEQ` is true if `cmp` is an
 * `==` operation and false if it is an `!=` operation.
 *
 * For example, if `rel` is `x == 5` then
 * `eqOpWithSwap(cmp, x, 5, true)` holds.
 */
private predicate eqOp(EqualityOperation cmp, Expr lhs, Expr rhs, boolean isEQ) {
  lhs = cmp.getLeftOperand().getFullyConverted() and
  rhs = cmp.getRightOperand().getFullyConverted() and
  (
    cmp instanceof EQExpr and isEQ = true
    or
    cmp instanceof NEExpr and isEQ = false
  )
}

/**
 * Holds if `cmp` is an equality operation (`==` or `!=`) with fully-converted
 * operands `a` and `b`, and `isEQ` is true if `cmp` is an `==` operation and
 * false if it is an `!=` operation.
 *
 * This allows for the equality to be either as written, or with its
 * arguments reversed; for example, if `cmp` is `x == 5` then both
 * `eqOpWithSwap(cmp, x, 5, true)` and
 * `eqOpWithSwap(cmp, 5, x, true)` hold.
 */
private predicate eqOpWithSwap(EqualityOperation cmp, Expr a, Expr b, boolean isEQ) {
  eqOp(cmp, a, b, isEQ) or
  eqOp(cmp, b, a, isEQ)
}

/**
 * Holds if `cmp` is an equality operation (`==` or `!=`) with fully-converted
 * children `a` and `b`, `isEQ` is true if `cmp` is an `==` operation and
 * false if it is an `!=` operation, and the result is `branch`.
 *
 * This allows for the comparison to be either as written, or with its
 * arguments reversed; for example, if `cmp` is `x == 5` then
 * `eqOpWithSwapAndNegate(cmp, x, 5, true, true)`,
 * `eqOpWithSwapAndNegate(cmp, 5, x, true, true)`,
 * `eqOpWithSwapAndNegate(cmp, x, 5, false, false)` and
 * `eqOpWithSwapAndNegate(cmp, 5, x, false, false)` hold.
 */
predicate eqOpWithSwapAndNegate(EqualityOperation cmp, Expr a, Expr b, boolean isEQ, boolean branch) {
  eqOpWithSwap(cmp, a, b, branch) and isEQ = true
  or
  eqOpWithSwap(cmp, a, b, branch.booleanNot()) and isEQ = false
}

/**
 * Holds if `cmp` is an unconverted conversion of `a` to a Boolean that
 * evaluates to `isEQ` iff `a` is 0.
 *
 * Note that `a` can be `cmp` itself or a conversion thereof.
 */
private predicate eqZero(Expr cmp, Expr a, boolean isEQ) {
  // The `!a` expression tests `a` equal to zero when `a` is a number converted
  // to a Boolean.
  isEQ = true and
  exists(Expr notOperand | notOperand = cmp.(NotExpr).getOperand().getFullyConverted() |
    // In C++ code there will be a BoolConversion in `!myInt`
    a = notOperand.(BoolConversion).getExpr()
    or
    // In C code there is no conversion since there was no bool type before C99
    a = notOperand and
    not a instanceof BoolConversion // avoid overlap with the case above
  )
  or
  // The `(bool)a` expression tests `a` NOT equal to zero when `a` is a number
  // converted to a Boolean. To avoid overlap with the case above, this case
  // excludes conversions that are right below a `!`.
  isEQ = false and
  linearAccess(cmp, _, _, _) and
  // This test for `isCondition` implies that `cmp` is unconverted and that the
  // parent of `cfg` is not a `NotExpr` -- the CFG doesn't do branching from
  // inside `NotExpr`.
  cmp.isCondition() and
  // The GNU two-operand conditional expression is not supported for the
  // purpose of guards, but the value of the conditional expression itself is
  // modeled in the range analysis.
  not exists(ConditionalExpr cond | cmp = cond.getCondition() and cond.isTwoOperand()) and
  (
    // In C++ code there will be a BoolConversion in `if (myInt)`
    a = cmp.getFullyConverted().(BoolConversion).getExpr()
    or
    // In C code there is no conversion since there was no bool type before C99
    a = cmp.getFullyConverted() and
    not a instanceof BoolConversion // avoid overlap with the case above
  )
}

/**
 * Holds if `branch` of `cmp` is taken when `a` compares `isEQ` to zero.
 *
 * Note that `a` can be `cmp` itself or a conversion thereof.
 */
predicate eqZeroWithNegate(Expr cmp, Expr a, boolean isEQ, boolean branch) {
  // The comparison for _equality_ to zero is on the `true` branch when `cmp`
  // compares equal to zero and on the `false` branch when `cmp` compares not
  // equal to zero.
  eqZero(cmp, a, branch) and isEQ = true
  or
  // The comparison for _inequality_ to zero is on the `false` branch when
  // `cmp` compares equal to zero and on the `true` branch when `cmp` compares
  // not equal to zero.
  eqZero(cmp, a, branch.booleanNot()) and isEQ = false
}

/**
 * Holds if `expr` is equivalent to `p*v + q`, where `p` is a non-zero
 * number. This takes into account the associativity, commutativity and
 * distributivity of arithmetic operations.
 */
predicate linearAccess(Expr expr, VariableAccess v, float p, float q) {
  // Exclude 0 and NaN.
  (p < 0 or p > 0) and
  linearAccessImpl(expr, v, p, q)
}

/**
 * Holds if `expr` is equivalent to `p*v + q`.
 * This takes into account the associativity, commutativity and
 * distributivity of arithmetic operations.
 */
private predicate linearAccessImpl(Expr expr, VariableAccess v, float p, float q) {
  // Base case
  expr = v and p = 1.0 and q = 0.0
  or
  expr.(ReferenceDereferenceExpr).getExpr() = v and p = 1.0 and q = 0.0
  or
  // a+(p*v+b) == p*v + (a+b)
  exists(AddExpr addExpr, float a, float b |
    addExpr.getLeftOperand().isConstant() and
    a = addExpr.getLeftOperand().getFullyConverted().getValue().toFloat() and
    linearAccess(addExpr.getRightOperand(), v, p, b) and
    expr = addExpr and
    q = a + b
  )
  or
  // (p*v+a)+b == p*v + (a+b)
  exists(AddExpr addExpr, float a, float b |
    addExpr.getRightOperand().isConstant() and
    b = addExpr.getRightOperand().getFullyConverted().getValue().toFloat() and
    linearAccess(addExpr.getLeftOperand(), v, p, a) and
    expr = addExpr and
    q = a + b
  )
  or
  // a-(m*v+b) == -m*v + (a-b)
  exists(SubExpr subExpr, float a, float b, float m |
    subExpr.getLeftOperand().isConstant() and
    a = subExpr.getLeftOperand().getFullyConverted().getValue().toFloat() and
    linearAccess(subExpr.getRightOperand(), v, m, b) and
    expr = subExpr and
    p = -m and
    q = a - b
  )
  or
  // (p*v+a)-b == p*v + (a-b)
  exists(SubExpr subExpr, float a, float b |
    subExpr.getRightOperand().isConstant() and
    b = subExpr.getRightOperand().getFullyConverted().getValue().toFloat() and
    linearAccess(subExpr.getLeftOperand(), v, p, a) and
    expr = subExpr and
    q = a - b
  )
  or
  // +(p*v+q) == p*v + q
  exists(UnaryPlusExpr unaryPlusExpr |
    linearAccess(unaryPlusExpr.getOperand().getFullyConverted(), v, p, q) and
    expr = unaryPlusExpr
  )
  or
  // (larger_type)(p*v+q) == p*v + q
  exists(Cast cast, ArithmeticType sourceType, ArithmeticType targetType |
    linearAccess(cast.getExpr(), v, p, q) and
    sourceType = cast.getExpr().getUnspecifiedType() and
    targetType = cast.getUnspecifiedType() and
    // This allows conversion between signed and unsigned, which is technically
    // lossy but common enough that we'll just have to assume the user knows
    // what they're doing.
    targetType.getSize() >= sourceType.getSize() and
    expr = cast
  )
  or
  // (p*v+q) == p*v + q
  exists(ParenthesisExpr paren |
    linearAccess(paren.getExpr(), v, p, q) and
    expr = paren
  )
  or
  // -(a*v+b) == -a*v + (-b)
  exists(UnaryMinusExpr unaryMinusExpr, float a, float b |
    linearAccess(unaryMinusExpr.getOperand().getFullyConverted(), v, a, b) and
    expr = unaryMinusExpr and
    p = -a and
    q = -b
  )
  or
  // m*(a*v+b) == (m*a)*v + (m*b)
  exists(MulExpr mulExpr, float a, float b, float m |
    mulExpr.getLeftOperand().isConstant() and
    m = mulExpr.getLeftOperand().getFullyConverted().getValue().toFloat() and
    linearAccess(mulExpr.getRightOperand(), v, a, b) and
    expr = mulExpr and
    p = m * a and
    q = m * b
  )
  or
  // (a*v+b)*m == (m*a)*v + (m*b)
  exists(MulExpr mulExpr, float a, float b, float m |
    mulExpr.getRightOperand().isConstant() and
    m = mulExpr.getRightOperand().getFullyConverted().getValue().toFloat() and
    linearAccess(mulExpr.getLeftOperand(), v, a, b) and
    expr = mulExpr and
    p = m * a and
    q = m * b
  )
}

/**
 * Holds if `guard` is a comparison operation (`<`, `<=`, `>`, `>=`,
 * `==` or `!=`), one of its arguments is equivalent (up to
 * associativity, commutativity and distributivity or the simple
 * arithmetic operations) to `p*v + q` (for some `p` and `q`),
 * `direction` describes whether `guard` give an upper or lower bound
 * on `v`, and `branch` indicates which control-flow branch this
 * bound is valid on.
 *
 * For example, if `guard` is `2*v + 3 > 10` then
 * `cmpWithLinearBound(guard, v, Greater(), true)` and
 * `cmpWithLinearBound(guard, v, Lesser(),  false)` hold.
 * If `guard` is `4 - v > 5` then
 * `cmpWithLinearBound(guard, v, Lesser(),  false)` and
 * `cmpWithLinearBound(guard, v, Greater(), true)` hold.
 *
 * A more sophisticated predicate, such as `boundFromGuard`, is needed
 * to compute an actual bound for `v`. This predicate can be used if
 * you just want to check whether a variable is bounded, or to restrict
 * a more expensive analysis to just guards that bound a variable.
 */
predicate cmpWithLinearBound(
  ComparisonOperation guard, VariableAccess v,
  RelationDirection direction, // Is this a lower or an upper bound?
  boolean branch // Which control-flow branch is this bound valid on?
) {
  exists(Expr lhs, float p, RelationDirection dir |
    linearAccess(lhs, v, p, _) and
    relOpWithSwapAndNegate(guard, lhs, _, dir, _, branch) and
    (
      p > 0 and direction = dir
      or
      p < 0 and direction = negateDirection(dir)
    )
  )
  or
  exists(Expr lhs |
    linearAccess(lhs, v, _, _) and
    eqOpWithSwap(guard, lhs, _, branch)
  )
}

/**
 * Holds if `lb` and `ub` are the lower and upper bounds of the unspecified
 * type `t`.
 *
 * For example, if `t` is a signed 32-bit type then holds if `lb` is
 * `-2^31` and `ub` is `2^31 - 1`.
 */
private predicate typeBounds(ArithmeticType t, float lb, float ub) {
  exists(IntegralType integralType, float limit |
    integralType = t and limit = 2.pow(8 * integralType.getSize())
  |
    if integralType instanceof BoolType
    then lb = 0 and ub = 1
    else
      if integralType.isSigned()
      then (
        lb = -(limit / 2) and ub = (limit / 2) - 1
      ) else (
        lb = 0 and ub = limit - 1
      )
  )
  or
  // This covers all floating point types. The range is (-Inf, +Inf).
  t instanceof FloatingPointType and lb = -(1.0 / 0.0) and ub = 1.0 / 0.0
}

private Type stripReference(Type t) {
  if t instanceof ReferenceType then result = t.(ReferenceType).getBaseType() else result = t
}

/** Gets the type used by range analysis for the given `StackVariable`. */
Type getVariableRangeType(StackVariable v) { result = stripReference(v.getUnspecifiedType()) }

/**
 * Gets the lower bound for the unspecified type `t`.
 *
 * For example, if `t` is a signed 32-bit type then the result is
 * `-2^31`.
 */
float typeLowerBound(Type t) { typeBounds(stripReference(t), result, _) }

/**
 * Gets the upper bound for the unspecified type `t`.
 *
 * For example, if `t` is a signed 32-bit type then the result is
 * `2^31 - 1`.
 */
float typeUpperBound(Type t) { typeBounds(stripReference(t), _, result) }

/**
 * Gets the minimum value that this expression could represent, based on
 * its type.
 *
 * For example, if `expr` has a signed 32-bit type then the result is
 * `-2^31`.
 *
 * Note: Due to the way casts are represented, rather than calling
 * `exprMinVal(expr)` you will normally want to call
 * `exprMinVal(expr.getFullyConverted())`.
 */
float exprMinVal(Expr expr) { result = typeLowerBound(expr.getUnspecifiedType()) }

/**
 * Gets the maximum value that this expression could represent, based on
 * its type.
 *
 * For example, if `expr` has a signed 32-bit type then the result is
 * `2^31 - 1`.
 *
 * Note: Due to the way casts are represented, rather than calling
 * `exprMaxVal(expr)` you will normally want to call
 * `exprMaxVal(expr.getFullyConverted())`.
 */
float exprMaxVal(Expr expr) { result = typeUpperBound(expr.getUnspecifiedType()) }

/**
 * Gets the minimum value that this variable could represent, based on
 * its type.
 *
 * For example, if `v` has a signed 32-bit type then the result is
 * `-2^31`.
 */
float varMinVal(Variable v) { result = typeLowerBound(v.getUnspecifiedType()) }

/**
 * Gets the maximum value that this variable could represent, based on
 * its type.
 *
 * For example, if `v` has a signed 32-bit type then the result is
 * `2^31 - 1`.
 */
float varMaxVal(Variable v) { result = typeUpperBound(v.getUnspecifiedType()) }
