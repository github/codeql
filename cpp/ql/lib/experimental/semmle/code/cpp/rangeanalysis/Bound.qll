import semmle.code.cpp.rangeanalysis.new.internal.semantic.analysis.Bound
