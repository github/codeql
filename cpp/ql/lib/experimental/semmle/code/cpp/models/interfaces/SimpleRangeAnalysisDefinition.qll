/**
 * EXPERIMENTAL: The API of this module may change without notice.
 *
 * Provides a class for modeling `RangeSsaDefinition`s with a restricted range.
 */

import cpp
import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis

/**
 * EXPERIMENTAL: The API of this class may change without notice.
 *
 * An SSA definition for which a range can be deduced. As with
 * `RangeSsaDefinition` and `SsaDefinition`, instances of this class
 * correspond to points in the program where one or more variables are defined
 * or have their value constrained in some way.
 *
 * Extend this class to add functionality to the range analysis library.
 */
abstract class SimpleRangeAnalysisDefinition extends RangeSsaDefinition {
  /**
   * Holds if this `SimpleRangeAnalysisDefinition` adds range information for
   * `v`. Because a `SimpleRangeAnalysisDefinition` is just a point in the
   * program, it's possible that more than one variable might be defined at
   * this point. This predicate clarifies which variable(s) should get range
   * information from `this`.
   *
   * This predicate **must be overridden** to hold for any `v` that can show
   * up in the other members of `SimpleRangeAnalysisDefinition`. Conversely,
   * the other members **must be accurate** for any `v` in this predicate.
   */
  abstract predicate hasRangeInformationFor(StackVariable v);

  /**
   * Holds if `(this, v)` depends on the range of the unconverted expression
   * `e`. This information is used to inform the range analysis about cyclic
   * dependencies. Without this information, range analysis might work for
   * simple cases but will go into infinite loops on complex code.
   *
   * For example, when modeling the definition by reference in a call to an
   * overloaded `operator=`, written as `v = e`, the definition of `(this, v)`
   * depends on `e`.
   */
  abstract predicate dependsOnExpr(StackVariable v, Expr e);

  /**
   * Gets the lower bound of the variable `v` defined by this definition.
   *
   * Implementations of this predicate should use
   * `getFullyConvertedLowerBounds` and `getFullyConvertedUpperBounds` for
   * recursive calls to get the bounds of their dependencies.
   */
  abstract float getLowerBounds(StackVariable v);

  /**
   * Gets the upper bound of the variable `v` defined by this definition.
   *
   * Implementations of this predicate should use
   * `getFullyConvertedLowerBounds` and `getFullyConvertedUpperBounds` for
   * recursive calls to get the bounds of their dependencies.
   */
  abstract float getUpperBounds(StackVariable v);
}

import SimpleRangeAnalysisInternal
