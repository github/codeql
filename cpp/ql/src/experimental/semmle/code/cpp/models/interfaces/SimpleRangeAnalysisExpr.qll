/**
 * EXPERIMENTAL: The API of this module may change without notice.
 *
 * Provides a class for modeling `Expr`s with a restricted range.
 */

import cpp
import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis

/**
 * EXPERIMENTAL: The API of this class may change without notice.
 *
 * An expression for which a range can be deduced. Extend this class to add
 * functionality to the range analysis library.
 */
abstract class SimpleRangeAnalysisExpr extends Expr {
  /**
   * Gets the lower bound of the expression.
   *
   * Implementations of this predicate should use
   * `getFullyConvertedLowerBounds` and `getFullyConvertedUpperBounds` for
   * recursive calls to get the bounds of their children.
   */
  abstract float getLowerBounds();

  /**
   * Gets the upper bound of the expression.
   *
   * Implementations of this predicate should use
   * `getFullyConvertedLowerBounds` and `getFullyConvertedUpperBounds` for
   * recursive calls to get the bounds of their children.
   */
  abstract float getUpperBounds();

  /**
   * Holds if the range this expression depends on the definition `srcDef` for
   * StackVariable `srcVar`.
   *
   * Because this predicate cannot be recursive, most implementations should
   * override `dependsOnChild` instead.
   */
  predicate dependsOnDef(RangeSsaDefinition srcDef, StackVariable srcVar) { none() }

  /**
   * Holds if this expression depends on the range of its unconverted
   * subexpression `child`. This information is used to inform the range
   * analysis about cyclic dependencies. Without this information, range
   * analysis might work for simple cases but will go into infinite loops on
   * complex code.
   *
   * For example, when modeling a function call whose return value depends on
   * all of its arguments, implement this predicate as
   * `child = this.getAnArgument()`.
   */
  abstract predicate dependsOnChild(Expr child);
}

import SimpleRangeAnalysisInternal

/**
 * This class exists to prevent the QL front end from emitting compile errors
 * inside `SimpleRangeAnalysis.qll` about certain conjuncts being empty
 * because the overrides of `SimpleRangeAnalysisExpr` that happen to be in
 * scope do not make use of every feature it offers.
 */
private class Empty extends SimpleRangeAnalysisExpr {
  Empty() {
    // This predicate is complicated enough that the QL type checker doesn't
    // see it as empty but simple enough that the optimizer should.
    this = this and none()
  }

  override float getLowerBounds() { none() }

  override float getUpperBounds() { none() }

  override predicate dependsOnChild(Expr child) { none() }
}
