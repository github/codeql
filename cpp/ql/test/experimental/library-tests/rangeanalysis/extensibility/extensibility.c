/// Adds its arguments (has custom modeling in QL)
int custom_add_function(int a, int b);

int test_extensibility_add(int x) {
  if (x >= -10 && x <= 10) {
    int result = custom_add_function(x, 100);
    return result; // 90 .. 110
  }
}

int test_overridability_sub(int x) {
  int result = x - (unsigned char)x; // Returns 0 due to custom modeling for this test being deliberately wrong
  return result; // 0
}

void test_parameter_override(int magic_name_at_most_10, int magic_name_at_most_20) {
  magic_name_at_most_10;
  magic_name_at_most_20;
}
