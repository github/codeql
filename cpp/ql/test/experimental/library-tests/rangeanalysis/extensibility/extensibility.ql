import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
import semmle.code.cpp.rangeanalysis.RangeAnalysisUtils
import experimental.semmle.code.cpp.models.interfaces.SimpleRangeAnalysisExpr
import experimental.semmle.code.cpp.models.interfaces.SimpleRangeAnalysisDefinition

class CustomAddFunctionCall extends SimpleRangeAnalysisExpr, FunctionCall {
  CustomAddFunctionCall() { this.getTarget().hasGlobalName("custom_add_function") }

  override float getLowerBounds() {
    exists(float lower0, float lower1 |
      lower0 = getFullyConvertedLowerBounds(this.getArgument(0)) and
      lower1 = getFullyConvertedLowerBounds(this.getArgument(1)) and
      // Note: this rounds toward 0, not -Inf as it should
      result = lower0 + lower1
    )
  }

  override float getUpperBounds() {
    exists(float upper0, float upper1 |
      upper0 = getFullyConvertedUpperBounds(this.getArgument(0)) and
      upper1 = getFullyConvertedUpperBounds(this.getArgument(1)) and
      // Note: this rounds toward 0, not Inf as it should
      result = upper0 + upper1
    )
  }

  override predicate dependsOnChild(Expr child) { child = this.getAnArgument() }
}

class SelfSub extends SimpleRangeAnalysisExpr, SubExpr {
  SelfSub() {
    this.getLeftOperand().(VariableAccess).getTarget() =
      this.getRightOperand().(VariableAccess).getTarget()
  }

  override float getLowerBounds() { result = 0 }

  override float getUpperBounds() { result = 0 }

  override predicate dependsOnChild(Expr child) { child = this.getAnOperand() }
}

/**
 * A definition for test purposes of a parameter `p` that starts with a
 * special prefix. This class is written to exploit how QL behaves when class
 * fields are not functionally determined by `this`. When multiple parameters
 * of the same function have the special prefix, there is still only one
 * instance of this class.
 */
class MagicParameterName extends SimpleRangeAnalysisDefinition {
  Parameter p;
  float value;

  MagicParameterName() {
    this.definedByParameter(p) and
    value = p.getName().regexpCapture("magic_name_at_most_(\\d+)", 1).toFloat()
  }

  override predicate hasRangeInformationFor(StackVariable v) { v = p }

  override predicate dependsOnExpr(StackVariable v, Expr e) {
    // No dependencies. This sample class yields constant values.
    none()
  }

  override float getLowerBounds(StackVariable var) {
    var = p and
    result = typeLowerBound(p.getUnspecifiedType())
  }

  override float getUpperBounds(StackVariable var) {
    var = p and
    result = value
  }
}

from VariableAccess expr, float lower, float upper
where
  lower = lowerBound(expr) and
  upper = upperBound(expr)
select expr, lower, upper
