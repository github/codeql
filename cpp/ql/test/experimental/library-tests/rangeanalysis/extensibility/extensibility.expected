| extensibility.c:5:7:5:7 | x | -2.147483648E9 | 2.147483647E9 |
| extensibility.c:5:19:5:19 | x | -10.0 | 2.147483647E9 |
| extensibility.c:6:38:6:38 | x | -10.0 | 10.0 |
| extensibility.c:7:12:7:17 | result | 90.0 | 110.0 |
| extensibility.c:12:16:12:16 | x | -2.147483648E9 | 2.147483647E9 |
| extensibility.c:12:35:12:35 | x | -2.147483648E9 | 2.147483647E9 |
| extensibility.c:13:10:13:15 | result | 0.0 | 0.0 |
| extensibility.c:17:3:17:23 | magic_name_at_most_10 | -2.147483648E9 | 10.0 |
| extensibility.c:18:3:18:23 | magic_name_at_most_20 | -2.147483648E9 | 20.0 |
